### PR TITLE
feat(compiler): custom resolvers

### PR DESCRIPTION
**Pull request by AleksandrSl**
## Summary

It's an example of having custom resolvers and what we have to do to support them. Ofc we have to make path based i18n support internally along with the cookies, without custom resolvers.

There are both changes on the user side (demo) and our side (in the compiler, because I removed the custom resolver part for simplicity until we have requests, but PR also demonstrates how we can support path based i18n, probably it can also be done a bit cleaner, and we can provide middleware configuration as exported function, so users don't have to do it, similar how next-intl does this. suprisingly they have a ton of code around these)

## Changes

- [Key change 1]
- [Key change 2]

## Testing

**Business logic tests added:**

- [ ] [Describe test 1 - what behavior it validates]
- [ ] [Describe test 2 - what edge case it covers]
- [ ] All tests pass locally

## Visuals

**Required for UI/UX changes:**

- [ ] Before/after screenshots attached
- [ ] Video demo for interactions (< 30s)

## Checklist

- [ ] Changeset added (if version bump needed)
- [ ] Tests cover business logic (not just happy path)
- [ ] No breaking changes (or documented below)

Closes #[issue-number]

## Summary by CodeRabbit

* **New Features**
  * Added path-based locale routing through URL segments (e.g., /en/, /es/, /de/).
  * Implemented automatic locale detection from browser language preferences.
  * Enabled custom locale resolvers for flexible locale-switching behavior.
  * Enhanced navigation with locale-aware link handling.

### CODE REVIEW

File: packages/new-compiler/src/index.ts [hunk 1]
```diff
@@ -1 +1,2 @@
 export type { PartialLingoConfig } from "./types";
+export type { LocaleCode } from "lingo.dev/spec";
```
Comment (AleksandrSl) [inline packages/new-compiler/src/index.ts:2]: Makes sense to reexport I think, because it's used in many places, and compiler could be the single dependency.

File: packages/new-compiler/src/virtual/code-generator.ts [hunk 2]
```diff
@@ -57,5 +71,9 @@ export function persistLocale(locale) {
     document.cookie = \`${cookieName}=\${locale}; path=/; max-age=${maxAge}\`;
   }
 }
+
+export function getLocalePathname(locale) {
+  return null; // Not used for cookie-based routing
+}
```
Comment (AleksandrSl) [inline packages/new-compiler/src/virtual/code-generator.ts:77]:
```suggestion
```

File: packages/new-compiler/src/virtual/code-generator.ts [hunk 3]
```diff
@@ -39,6 +46,13 @@ export const getServerLocale = createNextCookieLocaleResolver({ cookieConfig: ${
  * Exports getClientLocale() and persistLocale() functions
  */
 export function generateClientLocaleModule(config: LingoConfig): string {
+  if (config.localePersistence.type === "custom") {
+    // For custom resolvers, import from an abstract path that will be
+    // resolved by Turbopack's resolveAlias to the actual user file
+    return `export { getClientLocale, persistLocale, getLocalePathname } from '@lingo.dev/compiler/virtual/locale-client';`;
```
Comment (AleksandrSl) [inline packages/new-compiler/src/virtual/code-generator.ts]:
```suggestion
    return `export { getClientLocale, persistLocale } from '@lingo.dev/compiler/virtual/locale-client';`;
```

File: packages/new-compiler/src/react/shared/LingoProvider.tsx [hunk 4]
```diff
@@ -178,22 +179,26 @@ function LingoProvider__Prod({
...
-      // 2. Update local state for immediate UI feedback
+      // Update local state for immediate UI feedback
       setLocaleState(newLocale);
 
-      // 3a. Next.js pattern: Trigger server re-render
+      // Next.js pattern: Trigger server re-render
       if (router) {
-        router.refresh();
+        if (newUrl) {
+          router.push(newUrl);
```
Comment (AleksandrSl) [inline packages/new-compiler/src/react/shared/LingoProvider.tsx:196]: I missed that for the path based i18n we will have to do navigation. I don't think we need to add a separate method for that, just returning newUrl form `persistLocale` if redirect is needed sounds reasonable to me.

Next needs a `push`, because we have to trigger a middleware so that server components get a correct locale

File: packages/new-compiler/src/plugin/resolve-locale-resolver.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  sourceRoot: string,
+  lingoDir: string,
+  projectRoot: string,
+): {
+  serverResolver: string;
+  clientResolver: string;
+} {
+  const baseDir = path.join(projectRoot, sourceRoot, lingoDir);
+
+  const serverPath = resolveResolverPath("locale-resolver-server", baseDir);
```
Comment (AleksandrSl) [inline packages/new-compiler/src/plugin/resolve-locale-resolver.ts:69]: The whole file is a way to check that the file are present where we expect them to be. In this PR I expect them to be in the lingo directory, but we are free to come up with any pattern we want.

Also a nice error if one of the files is missing when custom locale resolver is chosen in settings would be good

File: packages/new-compiler/src/plugin/next.ts [hunk 6]
```diff
@@ -259,9 +260,31 @@ export async function withLingo(
...
+  //
+  // Convention: Resolver files must be at <sourceRoot>/<lingoDir>/locale-resolver-{server|client}.ts
+  if (lingoConfig.localePersistence.type === "custom") {
+    const resolvedPaths = resolveCustomResolverPaths(
+      lingoConfig.sourceRoot,
+      lingoConfig.lingoDir,
+      process.cwd(),
+    );
+
+    customResolverAliases = {
```
Comment (AleksandrSl) [inline packages/new-compiler/src/plugin/next.ts:279]: For vite and unplugin there should already be a similar thing.

See 
```
function tryLocalOrReturnVirtual(
  config: LingoConfig,
  fileName: string,
  virtualName: string,
) {
  const customPath = path.join(config.sourceRoot, config.lingoDir, fileName);
  if (fs.existsSync(customPath)) {
    return customPath;
  }
  return virtualName;
}
```

I don't remember if I checked it working

File: demo/new-compiler-next16/supported-locales.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+import type { LocaleCode } from "@lingo.dev/compiler"
```
Comment (AleksandrSl) [inline demo/new-compiler-next16/supported-locales.ts:1]: Just  a way to avoid repeating these

File: demo/new-compiler-next16/proxy.ts [hunk 8]
```diff
@@ -0,0 +1,107 @@
+import { NextRequest, NextResponse } from "next/server";
```
Comment (AleksandrSl) [inline demo/new-compiler-next16/proxy.ts:1]: It's 2026 why the hell Next asks users to configure all this hell when they want path based i18n
